### ISSUE TITLE
"Cannot convert null to … because it is a non-nullable value type" false positive

### ISSUE THREAD

**Issue by ForNeVeR** [dotnet/roslyn#52015]
**Version Used**: C# 9, .NET SDK 5.0.201

**Steps to Reproduce**:

1. `dotnet new console`
2. Write the following program:
   ```csharp
   namespace ConsoleApp3
   {
       public readonly struct MyStringValues
       {
           private readonly string _value;
           private readonly string[] _values;
   
           public MyStringValues(string value)
           {
               _value = value;
               _values = null;
           }
   
           public MyStringValues(string[] values)
           {
               _value = null;
               _values = values;
           }
           
           public static implicit operator MyStringValues(string value)
           {
               return new MyStringValues(value);
           }
   
           public static implicit operator MyStringValues(string[] values)
           {
               return new MyStringValues(values);
           }
   
           public static implicit operator string (MyStringValues values)
           {
               throw null;
           }
       }
       
       class Program
       {
           static void Main(string[] args)
           {
               var msv = new MyStringValues();
               var condition = true;
               string f = condition ? msv : null;
           }
       }
   }
   ```
   ([sharplab link](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4wmApgZwA4QDGWABAMID2MOFANlgIJ54DMAsAFADenJfJAARYkATlggIqtAJ4kcAFxEBXQvJIBZaQGVFsAOYA1CLSW5e/Hh37WSeEVABuEeaTESpsgQEYADCQD6TiZYANzmNrb2Ti6i4pIwMoK+ANoAugFBpjhhHOE2QhraujCGxlkAFN5+mVgAlHnWlhERgWWkALwkNTnNNq3BOCSdMEq0tD3NAL6cDfwFmjr2JUYDlSnpNTj1Vs1Nvfz9pkMkI2MT+xltg52b5zbTuTsRBd4AbCRQYHi0UIRQahQ8FgRM4KCJCot9CsKlUum1tr09hcBAB2E5YADuEOKpVWNVqd2sD1mfBeXnen2+v3+JEBwNB4IWOOhuDWPjScIGCN2JPyaJgmOxS1xFU2BN5xKe+WEbw+Xx+fwBQJB8jBSR85SZwpZgzFvKR+3kAAsRBQsadxhK8pLrAIAExJFF5A228mCAAsGmgMDZHIgIj0W31vOsTnBYBwDmOAqxWqhV3K4qlvTDJEIVAQ/ygVGOilMhOasIAZsd0zBM/JszASAB+EgRqMgE6jS3J/gPImcSZAA==))
3. Try to compile it

**Expected Behavior**:
This program should compile successfully.

**Actual Behavior**:
```
Program.cs(42,45): Error CS0037 : Cannot convert null to 'MyStringValues' because it is a non-nullable value type
```

Notably, the error will go away if I comment out the cast operator from string array to `MyStringValues`, which shouldn't have any relation to this at all (or at least I can't see how it relates to the result here):
```csharp
        public static implicit operator MyStringValues(string[] values)
        {
            return new MyStringValues(values);
        }
```
([sharplab link with commented-out operator](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4wmApgZwA4QDGWABAMID2MOFANlgIJ54DMAsAFADenJfJAARYkATlggIqtAJ4kcAFxEBXQvJIBZaQGVFsAOYA1CLSW5e/Hh37WSeEVABuEeaTESpsgQEYADCQD6TiZYANzmNrb2Ti6i4pIwMoK+ANoAugFBpjhhHOE2QhraujCGxlkAFN5+mVgAlHnWlhERgWWkALwkNTnNNq3BOCSdMEq0tD3NAL6cDfwFmjr2JUYDlSnpNTj1Vs1Nvfz9pkMkI2MT+xltg52b5zbTuTsRBd4AbCRQYHi0UIRQahQ8FgRM4KCJCot9CsKlUum1tr09hcBAB2E5YADuEOKpVWNVqd2sD1mfAA9KSXl53p9vr9/iRAcDQeCFjjobg1j40nCBgjmuSkTZydZUeisaylriKpsCSSSOTiU98sI3h8vj8/gCgSD5GCkj5yhKoVceVk+RFBb15AALEQULGncZyh5EmZKgQAJiSKLylrmVMEABYNNAYJzuRARHotnK/REnOCwDgHMcYJjsZL2ThyrKlb0EyRCFQEP8oFRjopTITmrCAGbHIswEvyMswEgAfhISZTIBOoydef4LqHnEmQA==))

**Comment by huoyaoyuan** [comment]
This should be a corner case of `?:` typing. To keep compatibility with old versions, target type will only be used when it has to be.
Implicit conversion can take part in common type decision.

**Comment by jaredpar** [comment]
Thanks for taking the time to report this behavior. Unfortunately though this is "By Design". 

The reason this fails is that there are multiple ways in which the `null` can convert to `MyStringValues`. There are two `implicit` operators to choose from here and there is no way to choose a better one. Hence there is no way to determine a common type between both arms of the conditional. 

The reason this works when you comment out one of the operators is that it removes the ambiguity. The code can also be made to compile by changing the input type to either of the operators to be a `struct` as that also removes the ambiguity.

**Comment by ForNeVeR** [comment]
Thanks for the explanation! Could we improve the diagnostic message, though?

If there's an operator ambiguity, I expect the compiler to clearly state that (probably even list the operators that created it).
